projects/blenderbot2/agents/blenderbot2.py

```diff
@@ -742,7 +751,7 @@ def _set_memory_decoder_vec(self, observation: Message) -> Message:
                     self.opt['memory_decoder_ignore_phrase'],
                     self.opt['memory_decoder_delimiter'],
                 )
-            if self.add_person_tokens:
+            if self.add_person_tokens and not self.opt['use_multiparty']:
```

Comment:
Does this mean we keep the persona tokens for the multiparty chat? If I have it correctly,`memory_decoder_input` is going into the memory decoder (am I right). Don't we need to remove all the persona tokens from it and then rematch them with the memories later?

Reply:
Yes, `memory_decoder_input` goes into the memory decoder.

So`memory_decoder_input` (which defaults to `observation['full_text']`) first gets tokenized and saved as `observation['memory_decoder_vec']`:
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/blenderbot2.py#L752-L757

Afterwards, `observation['mememory_decoder_vec]` is then passed to `self.memory_decoder.generate_memories`
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/modules.py#L359-L364

Which then goes through each example in the batch, 
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/sub_modules.py#L327-L337

calls `self._batch_generate`, which observes the input and adds a reply as the output
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/sub_modules.py#L140-L148

and finally calls `self. _extract_from_raw_memories`, which adds the persona tokens.
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/sub_modules.py#L365-L375

These are all done before the `generated_memoreis` get passed to the `MemoryDecoder` in `self.access_longterm_memory`:
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/modules.py#L407-L414

Finally, `access_longterm_memory` then takes care of writing the `generated_memories` (which comes from memory decoder's replies to `observation['full_text']`), along with `memory_vec ` from `observation['personas']`, to memory.
https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/modules.py#L588-L609

https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/modules.py#L639-L649

Reply:
So I think you're right, if `self._batch_generate` doesn't want persona tokens, we should remove them first and restore them later

https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/sub_modules.py#L335

I think this means I'll need to modify the teacher and remove speaker labels in `observation['text']`, add a `speakers=['A', 'B', 'C']` field in `observation`.

So the teacher `text` would change from this:

```
text:
    Knight: Hello
    Queen: Hi
    King:

label:
    Hola
```

to this:
```
text:
    Hello
    Hi

label:
    Hola

speakers:
    Knight
    Queen
    King
   
```

I will also need to extend a custom `Batch` class that includes `batch.speakers` (if we don't want to touch `TorchAgent`), and maybe also add `batch.speakers` the `_model_input`:

https://github.com/facebookresearch/ParlAI/blob/5f13a2d805131688aaf5253d4e29adc8b4ad6020/projects/blenderbot2/agents/blenderbot2.py#L873-L909

It seems like I won't need to vectorize the `speakers` in the `batch`, since I only need the raw text.

Thoughts?